### ISSUE TITLE
Mouse not locking in Wine/Proton programs in fullscreen or windowed

### ISSUE THREAD

**Issue by ardishco-the-great** [hyprwm/Hyprland#2376]
It seems like I've found a major bug that could push away a lot of people playing games on Hyprland, It's got to do with mouse locking. Every time I run a game under Wine or Proton, (might not be exclusive to Wine or Proton) after tabbing out for a while and tab back in, the mouse locking is broken. The cursor seems to be still moving. I have tested this under Sway to see if the same behaviour could be observed there and after my testing, I can say that this is exclusively Hyprland behaviour. This doesn't look like It's a Proton or Wine bug, I've tested a lot of games extensively under Sway and Hyprland and It looks like Hyprland has some problems to iron out 

Test games were:
Deep Rock Galactic (Proton 7.0-6)
Overwatch 2 (Latest Wine-GE as of 27-05-2023) (DD//MM//YEAR)

Steps to Reproduce:
>Open a game run with Proton or Wine
>Switch to another workspace for a few minutes, maybe 10 if you want to be sure
>Switch back to the game and have your mouse lock ruined

Expected Behaviour
>For the mouse not to lose it's mouse locking properties after 

Noted Outcome:
>Mouse locking breaking after switching to a different workspace

Possible related bugs on the issue tracker:
https://github.com/hyprwm/Hyprland/issues/1732 (This issue is not the same as my mouse doesn't escape into my other monitor all the time but rather it looks like I am locked between certain regions while in-game which is caused by the mouse getting stuck on the edges of the screen.

It is also worth noting that a similar thing happens on Waydroid while trying to run Roblox on it, though it might be a function on Waydroid not working, since there are issues about that open the Waydroid's Github page too. (Can't test because this stopped working for me, will update if I can test)

**Comment by ardishco-the-great** [comment]
Just observed the same behavior whenever the game works normally, the mouse moves in the background but it does not escape the window, It plays normally. Not a problem but it seems like it's related.

I will attach a video when I can get some footage, OBS is acting up a bit.

**Comment by ardishco-the-great** [comment]
See here : https://github.com/ValveSoftware/gamescope/issues/196#issuecomment-1531716290
Others users report fullscreen applications don't capture the mouse on Hyprland. I haven't tried a native program. 

Also, here is the log. I did quite a few things here so It's quite long. If a cleaner log would be better, let me know and I will log into Hyprland once again to ONLY open steam and launch a game.

**Comment by vaxerski** [comment]
It would be cool to know what you mean by broken. Is the cursor lock just gone or what

**Comment by ardishco-the-great** [comment]
Constraint is the right word here in that case. Here is a video with footage of what I mean: 

https://storage.cloudconvert.com/tasks/486dba08-25bb-4695-a286-ab41b039ccac/2023-05-27%2023-56-06.mp4?AWSAccessKeyId=cloudconvert-production&Expires=1685308861&Signature=jpl6NHtrRBjbzXv%2FpGB0aj1gM0Y%3D&response-content-disposition=inline%3B%20filename%3D%222023-05-27%2023-56-06.mp4%22&response-content-type=video%2Fmp4

Basically, the mouse itself still moves in the background no matter of if it looks like it's constraining me to a certain region like in the video. The teleporting around is also another thing to note, I forgot to mention it.

**Comment by vaxerski** [comment]
jesus christ your voice jumpscared me. Oddity oddity, I might look into the constraints code tomorrow, does sway have the same issue?

**Comment by ardishco-the-great** [comment]
Haha sorry about the voice. My mic is a little weird.

About Sway, no it did not. I will be testing it more on Sway to be even more sure but I have been using Sway for almost the whole day and left the game on to test if it would do the same thing, but it did not on multiple occasions.

**Comment by ardishco-the-great** [comment]
Okay, I have tried without my other monitor (the 1080p 60hz one) and the same behavior is present (tested with DRG once again)

And also tried TF2 and left it on for a while and that seems unaffected. I suspect there's a function being called via Proton that's acting weird on Hyprland.

**Comment by ecmatthee** [comment]
I am experiencing the same issue.

Hardware info:
2 Monitor Setup (Second monitor is disabled during testing)
GPU: AMD RX 6700 XT
Hyprland config: I can provide my Hyprland config file if needed but it is basically the auto-generated one + vim keybings and i disabled preserve_split (dwindle).

Unlike @ardishco-the-great my cursor lock doesn't work even without window switching. I can confirm sway does not have this issue. Even with setting mousewarp to "force" in wine (with regedit) the issue is still present.

Games tested:
+ CSGO (Steam): Linux Native => no issue
+ SWAT 4 Gold Edition (GOG): Tested on Wine => issue present
+ Fallout New Vegas (GOG): Tested on Wine => issue present
+ Bioshock 1 (GOG): Tested on Wine => issue present
+ Bioshock 2 (GOG): Tested on Wine => issue present
+ Borderlands 2 (Steam): Tested on Proton => issue present

@NotAShelf looking at your config file repo it seems you enable gamemode and disable several animations - perhaps this fixes the cursor capturing. Could you perhaps test if running your games without gamemode enabled produces the bug.

@vaxerski Did you manage to find anything?

**Comment by EvilBoi123** [comment]
Hello guys, and future people i found a workaround for this issue, not comfortable but it do makes gaming less painful. 

https://github.com/hyprwm/Hyprland/issues/1732#issuecomment-1582838104

**Comment by vaxerski** [comment]
> On another note, I ran into the same issue of my cursor travelling when it should be "locked" in Minecraft. As an example, when I close my inventory I expect that when I open my inventory back up my cursor should be centred. This doesn't happen, and the cursor appears to be where ever it would be had it been free to move while I was controlling the player's view. This also happens when I open the menu with escape.

Can't repro.

What if you remove glfw-wayland and use regular glfw? IIRC mc uses glfw no?

**Comment by ardishco-the-great** [comment]
I would like to add that this is not an NVIDIA problem either since the observed behavior is also seen on AMD GPUs. I have an Intel iGPU that I can test with, though I doubt it has anything to do with GPUs. 

Another thing I'd like to add is that I have tested Minecraft like @RetoranPetra here and I cannot observe the same behavior. There is something affecting this and the bug doesn't seem to be consistent so it makes it that much more to reproduce, it seems.
I have been experiencing this bug a whole lot less lately though It's still there.

**Comment by ardishco-the-great** [comment]
> No, the WLR_NO_HARDWARE_CURSORS envvar is unset on my system (running on an AMD GPU).

>I do have access to a NVIDIA (which I presume the envvar is mainly for) and an Intel iGPU. I will do some testing on both of them.

Actually, I have an AMD GPU and I use it because my cursor looks like crap.


>To add to those I have tested Deep Rock Galactic on Proton Experimental & Proton 7.0-6 and cannot reproduce the issue.
To that I say,
> the bug doesn't seem to be consistent so it makes it that much harder to reproduce, it seems.
This is very strange because I can't chalk it up to a fucked up wine or proton prefix because I've reset it a lot, can't chalk it up to hardware because It works on other WMs and DEs yet for everyone else different games (or programs that lock your mouse) seem to have the same bug...?

Did you follow my instructions mentioned in the original post? @ecmatthee 
>Open a game run with Proton or Wine
>Switch to another workspace for a few minutes, maybe 10 if you want to be sure
>Switch back to the game and have your mouse lock ruined
If so, then that's pretty strange but It's not surprising given what I mentioned above...

Also, I have found a little bit of a "workaround" for this problem.
1. unfullscreen the game
2. tile it
3. left and right click an bunch of times on your wallpaper or anywhere except a window in the same workspace as the game with this behaviour
4. do this for every time this occurs
This works for me but it's obviously not a fix it's simply a way to get around the problem for now. 

@RetoranPetra you might find the above helpful until this gets patched.

**Comment by vars1ty** [comment]
Experiencing the same issue. Had to enable mouse warp to get it "sorted", although it still escapes the window and goes to the other monitor.

**Comment by ardishco-the-great** [comment]
@vars1ty What is mouse warping and how do you enable it? I'd like to try.

**Comment by vars1ty** [comment]
@ardishco-the-great - Essentially doesn't let your mouse leave the focus of the fullscreen application, it's not perfect as you have to focus applications via workspaces on Hyprland, but it works.

For enabling it, in Bottles just go to Virtual Desktop and enable it.
For normal wine/proton, https://ubuntuforums.org/showthread.php?t=1603263

**Comment by MoonBurst** [comment]
Yeah, this one seems to apply as well with Lutris games. It can be mitigated a bit with gamescope, but at times my cam will rocket all the way down/up.

**Comment by SkyLeite** [comment]
> Yeah, this one seems to apply as well with Lutris games. It can be mitigated a bit with gamescope, but at times my cam will rocket all the way down/up.

This happens in Final Fantasy XIV as well, running with XIVLauncher (so not Lutris or gamescope). Camera shoots way up but only sometimes, regardless of mouse movement when I hold to drag the camera.

**Comment by SkyLeite** [comment]
I did some more testing and found another weird behavior. Sorry if this is long or the diagrams are stupid. I wanted to be as clear as possible.

Context: the game is Final Fantasy XIV, a tab targeted game where you can hold the right mouse button to move the camera around your character. I have a triple monitor set up, with two 1080p monitors at the top and one 32:9 monitor at the bottom, essentially making one big rectangle of pixels. [diagram below]

![image](https://github.com/hyprwm/Hyprland/assets/16887983/4a40db45-4f75-44dd-b645-01b064f057d3)

The issue: Sometimes when holding right click to drag the camera, it will act as if I just did an instant drag down, making the camera face the ground.

# Camera mechanics
When you want to move the camera, the following normally happens:

1. You hold right click at coordinates A:B
2. The cursor disappears
3. You move the mouse around
4. The camera moves
5. You let go of right click
6. The cursor re-appears at the A:B coordinates

# Weird behavior

![image](https://github.com/hyprwm/Hyprland/assets/16887983/e5c6ac7c-29d2-4255-bf58-e0678234ad6b)

When I hold right click in the position C, everything works as expected (besides the issue of the camera being dragged sometimes). I can move the camera as much as I want, and the cursor reappears as expected.

However, doing the same thing at the edge of screen 3 causes the cursor to appear in other monitors. For example, holding right click at position A (labeled in the following diagram) and moving the mouse UP causes the camera to move _and the cursor to appear in position B (monitor 1)_.

Doing the exact same motion with the cursor in position C (for a prolonged period of time) does not yield the same results. I can never see the cursor in other monitors. This leads me to believe this is a result of something weird in the code that handles moving the cursor through monitors.

**Comment by vaxerski** [comment]
github auto-closes but this needs testing with the mr

**Comment by ardishco-the-great** [comment]
I was away from my PC for a while. Like 3 weeks or something but coming back to this now, I can confirm this behaviour can be observed, though not as common. It's decreased with the recent updates. I don't know which addition or edit did it, but it is worth mentioning.

**Comment by vars1ty** [comment]
@vaxerski Since this was mentioned as fixed in rel. v0.28.0, perhaps the issue could be closed too?

**Comment by vaxerski** [comment]
idk if it's fixed, ardish claims its not

**Comment by SkyLeite** [comment]
Doesn't seem to be fixed for me either unfortunately :/

**Comment by fortnider** [comment]
Can confirm it's not fixed,

**Comment by ardishco-the-great** [comment]
> @vaxerski Since this was mentioned as fixed in rel. v0.28.0, perhaps the issue could be closed too?

Not entirely fixed yet, still can be observed and reproduced the same way as of right now.

**Comment by 3laws** [comment]
I can also vouch it's not fixed. I play Apex Legends that this has gotten me killed so many times I just give up after a while of ranked.

**Comment by anglakey** [comment]
This also occurs in both Age of Empires 2 and Dota 2, running both using gamescope.

**Comment by ardishco-the-great** [comment]
> > @vaxerski Since this was mentioned as fixed in rel. v0.28.0, perhaps the issue could be closed too?
> 
> Not entirely fixed yet, still can be observed and reproduced the same way as of right now.
> 
> I've tried more games than I can count as of now and can confirm this is indeed a problem with Hyprland.

Okay, I apologize for the lack of information here. I thought the Arch repos had the latest version and but it did not.

I have updated to the latest version of Hyprland and after doing some testing, It looks like this issue is fixed for me! Will report back if this happens again...

**Comment by anglakey** [comment]
Just to add, I am also on the latest version but still have issues.

**Comment by muni-corn** [comment]
same here; all between Valheim, Muck, and Minecraft. even Discord in activities like their mini golf game. i don't think this is limited to just wine/proton.

i was experiencing some strange behavior with Minecraft just now: i noticed that usually whenever i would go to press either left click or right click, the mouse would escape the window (fullscreen at the time) and jump to my other monitor. it looked to be centered on the monitor when it warps, too. this seems to happen regardless of what monitor the window is on.

come to think of it, i was experiencing this same issue with Discord. in a game of mini golf, i would try to click and drag with any mouse button and the cursor would warp to the other monitor.

sometimes, the mouse button up event doesn't get sent to the window, so i've had moments here and there in games where my character will be stuck doing a left- or right-click action (e.g. swinging an axe in minecraft) until a mouse-button-up event is sent again.

**Comment by ardishco-the-great** [comment]
> > > @vaxerski Since this was mentioned as fixed in rel. v0.28.0, perhaps the issue could be closed too?
> > 
> > 
> > Not entirely fixed yet, still can be observed and reproduced the same way as of right now.
> > I've tried more games than I can count as of now and can confirm this is indeed a problem with Hyprland.
> 
> Okay, I apologize for the lack of information here. I thought the Arch repos had the latest version and but it did not.
> 
> I have updated to the latest version of Hyprland and after doing some testing, It looks like this issue is fixed for me! Will report back if this happens again...

Okay. It happened again but it looks like this is mostly an XWayland and Wine thing. Wine plays weirdly with wine/proton on Wayland because it currently only supports X and therefore runs on XWayland. 

See here for an example; https://gitlab.freedesktop.org/xorg/xserver/-/issues/

This is basically caused by
1. The absence of a Wayland driver for Wine so programs can properly function at full capacity
2. XWayland not changing its way of mouse locking. Apparently, according to the code; the cursor is supposed to be hidden or else it won't mouse lock. This is a simplification of what the code in XWayland does (someone from the Grapejuice discord server told me about this)

I think a patch after Wine receives the Wayland driver update would be better since there are multiple things causing this. XWayland developers don't budge and write it off as a feature or that it's for security so there's no hope there. Mouse locking on Wayland will forever be a mess thanks to XWayland so we will just have to get Wayland native games.

**Comment by SkyLeite** [comment]
> Something is causing this to only happen on Hyprland as I've demonstrated but it's really hard to know what since Wine under XWayland is known to cause similar issues

If you compare this to Sway using League of Legends as an example, mouse locking works correctly in bordered windowed mode, but is extremely janky on windowed borderless mode. In Hyprland neither mode works.

**Comment by anglakey** [comment]
> Okay. It happened again but it looks like this is mostly an XWayland and Wine thing. Wine plays weirdly with wine/proton on Wayland because it currently only supports X and therefore runs on XWayland.
> 
> See here for an example; https://gitlab.freedesktop.org/xorg/xserver/-/issues/

> 1. The absence of a Wayland driver for Wine so programs can properly function at full capacity

Not sure how much of a wine issue this is, dota runs native in linux where the issue still occurs.

**Comment by ardishco-the-great** [comment]
> > Okay. It happened again but it looks like this is mostly an XWayland and Wine thing. Wine plays weirdly with wine/proton on Wayland because it currently only supports X and therefore runs on XWayland.
> > See here for an example; https://gitlab.freedesktop.org/xorg/xserver/-/issues/
> 
> > 1. The absence of a Wayland driver for Wine so programs can properly function at full capacity
> 
> Not sure how much of a wine issue this is, dota runs native in linux where the issue still occurs.

Yeah...That's why it's kind of hard to narrow down since XWayland and Wine running under it are both causing issues on top of something with Hyprland...

**Comment by anglakey** [comment]
It seems that this has been fixed as of v0.30.0

**Comment by vaxerski** [comment]
closing then

**Comment by ardishco-the-great** [comment]
I'm still waiting on nixpkgs to update so I can test this myself.

**Comment by ardishco-the-great** [comment]
Just updated to 0.30
![image](https://github.com/hyprwm/Hyprland/assets/109692107/260b8e4d-6da4-4957-9761-b462e7457752)

still present

tested on Payday 3, seems like this might go on for a while.

**Comment by EvilBoi123** [comment]
as of version 3.0, genshin impact now locks mouse. yay :)

**Comment by gabeyyplx** [comment]
Can confirm the cursor is properly constrained in League of Legends on Arch as of now 😼

**Comment by muni-corn** [comment]
Minecraft seems to be working properly now ^-^

**Comment by jennydaman** [comment]
> as of version 3.0, genshin impact now locks mouse. yay :)

Are you playing in fullscreen, windowed, or borderless windowed?

**Comment by EvilBoi123** [comment]
> Are you playing in fullscreen, windowed, or borderless windowed?

I play in fullscreen but i believe borderless window also works. 

> It's not working for me, borderless window mode, dual monitor setup, Hyprland v0.31.0

It works for me in dual monitor setup, it only works when you in "controlling character" mode where there will be no cursor render. But if when the cursor render such as when "settings, pause, backpack mode" the mouse will not lock.

**Comment by jennydaman** [comment]
@EvilBoi123  that would be the desirable behavior.


In “controlling character” mode, I believe the game tries to “lock” the cursor to the center of the screen (that’s what happens on Windows). When I play Genshin on Hyprland, if I move my mouse too far in the direction of my second monitor, the cursor escapes and the in-game camera stops panning. Cursor locking seems to work on GNOME.

**Comment by Chedski** [comment]
I'm having this issue with Wine/Xwayland, but _only_ when the mouse is visible while being locked (for example, Roblox's first person view and rotating the camera, which keeps the standard mouse visible)

**Comment by Hellscaped** [comment]
having this issue

**Comment by andmydignity** [comment]
I don't know about you guys, but for me the problem is the the refresh rate. If i use 100HZ it won't lock but if i reduce to 60HZ it will lock properly.

**Comment by catfish-k** [comment]
I am not sure if this is the same issue, but I had the problem where opening some games the mouse focus would retreat to the top left corner and not lock on the center of the screen in first/3rd person games, making the game unplayable.  A restart of hyprland would fix it every time.  This setting seems to work right now:

```
input {

    follow_mouse = 1
    mouse_refocus = false #<-- add this

}
```
Now if it happens, all I need to do is swap to another workspace then back and my mouse is locked into the game correctly.  I hope it helps!

**Comment by DJXJD** [comment]
I started having this issue as well 1 week ago after updating from 0.34.0 to 0.35.0.
In particular, I notice this behavior in cs2 and in smite, but not in overwatch 2. In cs, it very sadly happens most frequently right when an enemy appears on my screen...

If there's anything I can do to help solve this issue, please let me know and I'll do my best.

**Comment by DJXJD** [comment]
More info, this issue only seems to occur when I have a window open on my second monitor (usually discord). So far, this issue hasn't occurred when an empty workspace is open on my second monitor.

**Comment by SteavenGamerYT** [comment]
its not fixed

**Comment by vaxerski** [comment]
please check #4889

**Comment by krakow10** [comment]
I experienced an invisible mouse escaping window borders while playing DRG via proton on the center monitor with 3 monitors, it could escape to either side and would change keyboard focus on click if the invisible mouse was outside the game window, but would not change mouse focus.  My workaround was to put empty workspaces on the other screens.

> I'm having this issue as well, but _only_ when the mouse is visible while being locked (for example, Roblox's first person view and rotating the camera, which keeps the standard mouse visible)

Roblox mouse locking was (Roblox blocks wine now) an issue unique to Roblox which happened on any wayland DE, not just hyprland.

**Comment by prismatica-dev** [comment]
This issue was the reason I stopped using Hyprland about 6 months ago. It made almost every game unplayable. 
A few days ago I once again did a fresh install of EndeavousOS, installed Hyprland (0.35.0-2 from AUR) and was annoyed to see that the bug was still present. In Minecraft clicking anywhere would focus a program on my other monitor. 

**Now after switching to hyprland-git I can confirm that, at least in Minecraft, the issue is finally no longer present for me.** Seriously, thank you for fixing this.

Some additional info in case its relevant,

**Hyprland Packages:**
- hyprland-git 0.36.0.r106.893c5521-1
- hyprcursor-git 0.1.4.r1.g1761f6ce-1
- hyprlang 0.4.2-1
- xdg-desktop-portal-hyprland 1.3.1-5

**Monitor Configuration:**
- monitor=HDMI-A-1,1920x1080@60,0x0,1
- monitor=DP-1,1920x1080@60,1920x0,1

I cannot currently test the other games I remember being impacted by this issue due to issues with Steam, but I will edit this when I do.

**Comment by ItzDerock** [comment]
I have this issue when I switch windows out of THE FINALS, and then try to go back to the game. The cursor does not lock so if I turn in the game, my cursor will go to my other screen and the game will lose focus. Very annoying, re-fullscreening the game does not fix the issue (tried both in-game settings and hyprland keybinds that dispatch fullscreen verb). The mouse is invisible in-game when this occurs.

**System:**
- Hyprland built at commit 893c55217b15cf0e5fe7eb3eb2d88183df410b9a
- NVIDIA 545.29.02

Is there some sort of lock mouse bind that I can set so I can manually lock my mouse as a workaround? I may also give gamescope a try and will edit this message with the results.

**Comment by kensad-ops** [comment]
is there any way to fix this at the moment ? i have encounter this not just mouse but also keyboard (more rarely) ? at the moment i just pray to the machine god that my mouse or keyboard don't break after i change focus, workspace. Some time i was able to play fine, sometime it just break right when i start the game.

**Comment by vaxerski** [comment]
generally it's mostly problems with the apps themselves, but you can always add 1px of an offset to your monitor whenever gaming and that will prevent the mouse from escaping, I guess.

**Comment by vaxerski** [comment]
if it's an xwayland issue, then yea, closing

**Comment by XorvatoAmigos** [comment]
To be honest I thought the problem was in my configuration, the cursor seemed to lock into the main window, but randomly it goes out when making very quick movements, especially when pressing a large key combination, when sending the position of my main monitor into the stratosphere, none of this happens. 
For those who will be re-reading all of this, I'd like to leave _“Simplifying Life to a Beautiful Future without XWayland”._

This script, as reported above, changes the position of your main monitor, but now automatically. All you have to **change is the ID of your monitor**.

```bash
#!/usr/bin/env bash

state="/tmp/togglemonitorlock"
status="false"

# Specify the “ID” of the monitor that is used as the primary monitor, see using hyprctl monitors
id=0

# Monitor ID
monitor=$(hyprctl -j monitors | jq -r ".[] | select (.id == $id) | .name")

# Default position of the monitor
defaultPositionCmd="wlr-randr --json | jq \".[] | select (.name == \\\"${monitor}\\\") | .position\" | awk -F': ' '{print \$2}' | tr -d ' ' | tr -d '\n'"
defaultPosition=$(eval ${defaultPositionCmd})

# Current workspace ID
currentWorkspace=$(hyprctl -j activeworkspace | jq '.id')

# Checking the existence of the status file and reading values from it
if [[ -f ${state} ]]; then
    mapfile -t stateValues < ${state}
    defaultPosition="${stateValues[0]}"
    status="${stateValues[1]}"
fi

# Switching the monitor position and updating the status file
if [[ ${status} == "true" ]]; then
    echo ${defaultPosition} > ${state}
    wlr-randr --output ${monitor} --pos ${defaultPosition}
    echo "false" >> ${state}
else
    defaultPosition2=$(eval ${defaultPositionCmd})
    # Save the new monitor position if it has been changed by the user
    if [[ ${defaultPosition2} == ${defaultPosition} ]]; then
        echo ${defaultPosition} > ${state}
    else
        echo ${defaultPosition2} > ${state}
    fi
    wlr-randr --output ${monitor} --pos 2000,2000
    echo "true" >> ${state}
    sleep 0.3
    # Restoring workspace
    hyprctl dispatch workspace ${currentWorkspace}
fi

```

I would also like to thank vaxerski for a great window manager!

**Comment by catfish-k** [comment]
Warframe and Deep Rock Galactic have been running great for me lately, but Alt+Tab or switching workspace will mess up the mouse focus.  In Warframe, and only after switching work spaces, the mouse cursor works as intended until I click on something, at which point the cursor warps to the bottom right corner.  In DRG the mouse cursor works as intended for a few seconds then it warps away.

I recently switched from Nvidia graphics to an AMD card and, weirdly, workspace or Alt+Tab switching did work when I had Nvidia.  However the mouse locking to the top left and 3rd person spinning around bug was present occasionally on Nvidia, especially in Warframe.  Generally Nvidia ran piss poor and AMD is so much better, save this bug.

**Comment by raidenovich** [comment]
I'm gonna add more to say that this is an XWayland bug as I was able to reproduce this issue in native linux programs aswell (though on rarer occasions)

**Comment by Arcitec** [comment]
@ardishko Thank you for the information about GNOME. In fact, I found this ticket because I am here with the same issue on GNOME Wayland.

The issues I have are:

- Right clicking makes the cursor jump to a random spot on screen.
- Right clicking and dragging makes the real cursor move when looking around inside the game. Awful.

I use Wine 9.22 and tried everything people were talking about trying:

- Enable virtual desktop in Wine.
- Enable mouse warp in Wine.

Neither worked.

But THIS worked for me:

- Enable Gamescope, set it to fullscreen with my screen resolution (3840x2160) as both its game resolution and window resolution.
- Set the game (in my case World of Warcraft WotLK, the old game client) to fullscreen at 3840x2160.
- Now the game behaves exactly like on Windows.
- Thank you Gamescope!

**Comment by ardishko** [comment]
Wine 9.22 has the wayland backend enabled by default now iirc, def check that out if you haven't, It's most likely fixed everything. @Arcitec

**Comment by Arcitec** [comment]
@ardishko I was using Wine 9.22 and it still had the mouse jump issue. Not sure if the wayland backend was active or not though... I just used the Kron4ek Wine 9.22 Staging TKG variant in Bottles and it was awful. :)

**Comment by Arcitec** [comment]
I have now found out why Gamescope + Wayland + Any game with Vsync crashes.

It's an NVIDIA issue, and it's caused by a vsync API that has loooong been buggy. Cubanismo is an NVIDIA developer and comments about it here:

https://github.com/ValveSoftware/gamescope/issues/1592

If the game uses Direct3D 12 or newer (VKD3D), you can use `VKD3D_DISABLE_EXTENSIONS=VK_KHR_present_wait`.

If it uses Direct3D 11 or older (DXVK) you can't do anything except forcing the game to disable Vsync. For example by launching it without Gamescope, disabling Vsync, and then launching it with Gamescope.

**Comment by Arcitec** [comment]
I guess I am going back to X11 to play this game...

**Comment by Ein420** [comment]
I'm also experiencing this issue now, haven't before, specifically with the game Lethal Company on Steam, it only grabs the mouse on my second monitor (HDMI-A-1) during fullscreen, whilst on my main one it doesn't grab the mouse at all, please help

**Comment by catfish-k** [comment]
I have been having better luck lately.  The only game that doesn't lock my mouse, is Warframe.  However, if I switch Hyprland work spaces then back in, the mouse locks.

AMD RX6800
Intel i7-9700k
Arch Linux

Possible relevant Hyprland configs:

   _monitor=DP-1,2560x1440@165,0x0,1

   follow_mouse = 1

    default_monitor = DP-1
    hide_on_key_press = true
    inactive_timeout = 5
    no_hardware_cursors = true
    no_warps = true

   explicit_sync = 2
   explicit_sync_kms = 2
   direct_scanout = true
   expand_undersized_textures = true

    allow_tearing = true

   vrr = 1

    animate_manual_resizes = false
    focus_on_activate = true
    render_ahead_of_time = false

    new_status = master

    workspace_swipe = off_


My games are all configured VSYNC disabled and set for 165 hz fullscreen.

Another curious thing that happens is after some time of Hyprland running- many hours, my frame rates drop to around 30 fps (from ~160) until i log out of Hyprland and back in.

**Comment by RambilE** [comment]
> # Edit: Well, after trying Wine 10 and all the settings, I can conclude: Gamescope is STILL the ONLY way to get correct mouse capturing / locking on Wayland.

Are you running gamescope on flatpak steam or not? Because I'm facing the same issue that my cursor doesn't really work in proton game (it still moves when i move the mouse, the game simply doesn't know where the cursor is) and gamescope errors out with
```
[gamescope] [Error] xdg_backend: Couldn't create Wayland input objects.
[gamescope] [Error] xdg_backend: Failed to initialize input thread
```
And i can't find any fix for this, i've only found an issue that's about year old at this point

**Comment by FrihtzCatz** [comment]
> It seems like I've found a major bug that could push away a lot of people playing games on Hyprland, It's got to do with mouse locking. Every time I run a game under Wine or Proton, (might not be exclusive to Wine or Proton) after tabbing out for a while and tab back in, the mouse locking is broken. The cursor seems to be still moving. I have tested this under Sway to see if the same behaviour could be observed there and after my testing, I can say that this is exclusively Hyprland behaviour. This doesn't look like It's a Proton or Wine bug, I've tested a lot of games extensively under Sway and Hyprland and It looks like Hyprland has some problems to iron out
> 
> Test games were: Deep Rock Galactic (Proton 7.0-6) Overwatch 2 (Latest Wine-GE as of 27-05-2023) (DD//MM//YEAR)
> 
> Steps to Reproduce:
> 
> > Open a game run with Proton or Wine
> > Switch to another workspace for a few minutes, maybe 10 if you want to be sure
> > Switch back to the game and have your mouse lock ruined
> 
> Expected Behaviour
> 
> > For the mouse not to lose it's mouse locking properties after
> 
> Noted Outcome:
> 
> > Mouse locking breaking after switching to a different workspace
> 
> Possible related bugs on the issue tracker: [#1732](https://github.com/hyprwm/Hyprland/issues/1732) (This issue is not the same as my mouse doesn't escape into my other monitor all the time but rather it looks like I am locked between certain regions while in-game which is caused by the mouse getting stuck on the edges of the screen.
> 
> It is also worth noting that a similar thing happens on Waydroid while trying to run Roblox on it, though it might be a function on Waydroid not working, since there are issues about that open the Waydroid's Github page too. (Can't test because this stopped working for me, will update if I can test)

You have to set no hardware cursors settings in your hyprland.conf file. Took me about 8 hrs of hard googling and forum scraping to find this fix and has fixed it for all my games so far.

`
cursor {
	no_hardware_cursors = true
	}

`

**Comment by FrihtzCatz** [comment]
> > > # Edit: Well, after trying Wine 10 and all the settings, I can conclude: Gamescope is STILL the ONLY way to get correct mouse capturing / locking on Wayland.
> > 
> > 
> > Are you running gamescope on flatpak steam or not? Because I'm facing the same issue that my cursor doesn't really work in proton game (it still moves when i move the mouse, the game simply doesn't know where the cursor is) and gamescope errors out with
> > ```
> > [gamescope] [Error] xdg_backend: Couldn't create Wayland input objects.
> > [gamescope] [Error] xdg_backend: Failed to initialize input thread
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And i can't find any fix for this, i've only found an issue that's about year old at this point
> 
> Okay nevermind there are absolutely no issues when i run protonge10-3 with PROTON_ENABLE_WAYLAND=1

You have to add the no hardware cursor setting in your hyprland.conf file:

`
cursor {
	no_hardware_cursors = true
	}

`
